server/src/main/java/org/elasticsearch/common/compress/Compressor.java

```diff
@@ -32,11 +32,19 @@
 
     int headerLength();
 
+    /**
+     * Creates a new stream input that decompresses the contents read from teh provided stream input.
+     * Closing the returned {@link StreamInput} will close the provided stream input.
+     * Note: The returned stream may only be used on the thread that created it as it might use thread-local resources and must be safely
+     * closed after use
+     */
     StreamInput streamInput(StreamInput in) throws IOException;
 
     /**
      * Creates a new stream output that compresses the contents and writes to the provided stream
      * output. Closing the returned {@link StreamOutput} will close the provided stream output.
+     * Note: The returned stream may only be used on the thread that created it as it might use thread-local resources and must be safely
+     * closed after use
      */
     StreamOutput streamOutput(OutputStream out) throws IOException;
```

Comment:
Maybe name this method something like `unsafeStreamOutput` or `threadLocalStreamOutput`?

Reply:
Looking at your HTTP exporter comment here as well. That's actually fine I think because we invoke the `doFlush` on the exporter from a HTTP client IO dispatcher thread. That said ... that's not something where I'm 100% confident in doing this thread-local magic + it's totally unnecessary there. Maybe we could instead of renaming add a parameter `boolean threadLocal` here to indicate whether or not we want a thread-local stream, wdyt? :)